firestore/bulkwriter.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	if doc == nil {
+		return errors.New("firestore: nil document contents")
+	}
+
+	_, havePath := bw.docUpdatePaths[doc.shortPath]
+	if havePath {
+		return fmt.Errorf("firestore: bulkwriter: received duplicate write for path: %v", doc.shortPath)
+	}
+
+	bw.docUpdatePaths[doc.shortPath] = true
+
+	return nil
+}
+
+// write packages up write requests into bulkWriterJob objects.
+func (bw *BulkWriter) write(w *pb.Write) bulkWriterOperation {
+	r := make(chan *pb.WriteResult, 1)
+	e := make(chan error, 1)
+
+	j := bulkWriterOperation{
+		result: r,
+		write:  w,
+		err:    e,
+		size:   0, // ignore operation size constraints; can't be inferred at compile time
+	}
+
+	err := bw.bundler.Add(j, j.size)
+	if err != nil {
+		j.err <- err
+		j.result <- nil
+	}
+	return j
+}
+
+// processResults checks for errors returned from send() and packages up the
+// results as WriteResult objects
+func processResults(o bulkWriterOperation) (*WriteResult, error) {
+	wpb := <-o.result
+	err := <-o.err
+
+	if err != nil {
+		return nil, err
+	}
+
+	wr, err := writeResultFromProto(wpb)
+	if err != nil {
+		return nil, err
+	}
+	return wr, nil
+}
+
+// send transmits writes to the service and matches response results to job channels.
+func (bw *BulkWriter) send(i interface{}) {
+	bwj := i.([]bulkWriterOperation)
+
+	var ws []*pb.Write
+	for _, w := range bwj {
+		ws = append(ws, w.write)
+	}
+
+	bwr := pb.BatchWriteRequest{
+		Database: bw.database,
+		Writes:   ws,
+		Labels:   map[string]string{},
+	}
+
+	select {
+	case <-bw.ctx.Done():
+		break
+	default:
+		resp, err := bw.vc.BatchWrite(bw.ctx, &bwr)
+		if err != nil {
+			// Do we need to be selective about what kind of errors we send?
+			for _, j := range bwj {
+				j.result <- nil
+				j.err <- err
+			}
+			return
+		}
+		// Iterate over the response. Match successful requests with unsuccessful
+		// requests.
+		for i, res := range resp.WriteResults {
+			s := resp.Status[i]
+
+			c := s.GetCode()
+
+			if c != 0 { // Should we do an explicit check against rpc.Code enum?
+				bw.retryBundler.Add(bwj[i], bwj[i].size)
+				continue
+			}
+
+			bwj[i].result <- res
+			bwj[i].err <- nil
+		}
+		// This means the writes are now finalized, all retries completed
+	}
+
+	// Check whether we need to increase the rate of requests to the service
+	// after each result
+	bw.rateLock.Lock()
+	bw.increaseRate()
+	bw.rateLock.Unlock()
+}
+
+// increaseRate updates the number of bundler requests that can be concurrently open
+func (bw *BulkWriter) increaseRate() {
+	elapsed := time.Now().Sub(bw.start).Seconds()
+
+	// Ideally, we would determine the qps before increasing, but instead
+	// we simply increase the number of bundler handler invocations we have
+	// running at once.
+	mins := elapsed / 60
+	if math.Mod(mins, 5) <= 0.1 {
+		newOps := float64(bw.maxOpsPerSecond) * rateLimiterMultiplier
+		bw.maxOpsPerSecond = int(newOps)
+		bw.bundler.HandlerLimit = int(newOps)
```

Comment:
I am not sure this can be dynmaically changed. Would need testing. I suspect this is racy though because I don't think it is meant to be changed, so no mutexs